File: c++/day04.cpp
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

std::vector<std::string> readInput() {
  std::vector<std::string> input;
  std::ifstream inputStream("../inputs/04.txt");
  std::string line;

  while (std::getline(inputStream, line)) {
    input.push_back(line);
  }

  return input;
}

int checkXmas(int row, int col, const std::vector<std::string> &source) {
  std::string search = "XMAS";
  std::pair<int, int> directions[8] = {{-1, -1}, {-1, 0}, {-1, 1}, {0, -1},
                                       {0, 1},   {1, -1}, {1, 0},  {1, 1}};

  int sum = 0;
  for (int i = 0; i < 8; i++) {
    int dx = directions[i].first;
    int dy = directions[i].second;

    for (int letter_index = 1; letter_index < search.size(); letter_index++) {
      int newCol = col + letter_index * dx;
      int newRow = row + letter_index * dy;
      if (newCol < 0 || newCol >= source[0].size()) {
        break;
      }
      if (newRow < 0 || newRow >= source.size()) {
        break;
      }
      if (source[newRow][newCol] != search[letter_index]) {
        break;
      }
      if (letter_index == search.size() - 1) {
        sum++;
      }
    }
  }

  return sum;
}

bool checkMas(int row, int col, const std::vector<std::string> &source) {
  std::pair<int, int> directions[] = {{1, 1}, {-1, -1}, {1, -1}, {-1, 1}};
  std::pair<int, int> opposite[] = {{-1, -1}, {1, 1}, {-1, 1}, {1, -1}};

  for (int i = 0; i < 4; i++) {
    int newRow = row + directions[i].first;
    int newCol = col + directions[i].second;
    int newRowOpposite = row + opposite[i].first;
    int newColOpposite = col + opposite[i].second;
    char one = source[newRow][newCol];
    char two = source[newRowOpposite][newColOpposite];
    if (!((one == 'M' && two == 'S') || (one == 'S' && two == 'M'))) {
      return false;
    }
  }
  return true;
}

int silver() {
  auto input = readInput();
  int sum = 0;
  for (int row = 0; row < input.size(); row++) {
    for (int col = 0; col < input[0].size(); col++) {
      if (input[row][col] == 'X') {
        sum += checkXmas(row, col, input);
      }
    }
  }
  return sum;
}

int gold() {
  auto input = readInput();
  int sum = 0;
  for (int row = 1; row < input.size() - 1; row++) {
    for (int col = 1; col < input[0].size() - 1; col++) {
      if (input[row][col] == 'A') {
        sum += checkMas(row, col, input);
      }
    }
  }
  return sum;
}

int main() {
  std::cout << "silver: " << silver() << "\n";
  std::cout << "gold: " << gold() << "\n";
}

File: rust/src/day04.rs
use std::fs::read_to_string;

pub fn run() {
    let input = read_input("../inputs/04.txt");
    println!("silver: {}", silver(&input));
    println!("gold: {}", gold(&input));
}

fn silver(input: &Vec<String>) -> i32 {
    let mut sum = 0;
    for row in 0..input.len() {
        for (col, ch) in input[row].chars().enumerate() {
            if ch == 'X' {
                sum += check_xmas((row as i32, col as i32), input);
            }
        }
    }
    sum
}

fn gold(input: &Vec<String>) -> i32 {
    let mut sum = 0;
    for row in 1..input.len() - 1 {
        for (col, ch) in input[row].chars().enumerate() {
            if ch == 'A' {
                sum += check_mas((row as i32, col as i32), input) as i32;
            }
        }
    }
    sum
}

fn check_mas(pos: (i32, i32), source: &Vec<String>) -> bool {
    let directions = [(1, 1), (-1, -1), (1, -1), (-1, 1)];
    let opposite = [(-1, -1), (1, 1), (-1, 1), (1, -1)];

    for i in 0..directions.len() {
        let new_pos = (pos.0 + directions[i].0, pos.1 + directions[i].1);
        let pos_opposite = (pos.0 + opposite[i].0, pos.1 + opposite[i].1);
        let one = match source[new_pos.0 as usize].chars().nth(new_pos.1 as usize) {
            Some(n) => n,
            None => {
                return false;
            }
        };

        let two = match source[pos_opposite.0 as usize]
            .chars()
            .nth(pos_opposite.1 as usize)
        {
            Some(n) => n,
            None => {
                return false;
            }
        };

        if !((one == 'M' && two == 'S') || (one == 'S' && two == 'M')) {
            return false;
        }
    }
    true
}

fn check_xmas(pos: (i32, i32), source: &Vec<String>) -> i32 {
    let mut sum = 0;
    let search_string = "XMAS";
    let directions: [(i32, i32); 8] = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];

    for d in directions {
        for letter_idx in 1..search_string.len() {
            let new_pos = (
                pos.0 + (letter_idx as i32) * d.0,
                pos.1 + (letter_idx as i32) * d.1,
            );
            if new_pos.0 < 0
                || new_pos.0 >= source.len() as i32
                || new_pos.1 < 0
                || new_pos.1 >= source[0].len() as i32
            {
                break;
            }
            if source[new_pos.0 as usize].chars().nth(new_pos.1 as usize)
                != search_string.chars().nth(letter_idx)
            {
                break;
            }
            if letter_idx == search_string.len() - 1 {
                sum += 1;
            }
        }
    }
    sum
}

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .expect("Failed to read file")
        .lines()
        .map(String::from)
        .collect()
}
